### DISCUSSION TITLE
Feature Request: caching bitmap fonts in client?

### DISCUSSION THREAD

**Discussion by reececomo** [pixijs/pixijs#10354] [Q&A]
### Description

Is there any happy path for caching bitmap fonts? We pre-generate ~20 bitmapfonts for various styles during preload and it would be awesome to persist those

Alternative is we can manually create a bunch of .fnt files - but would prefer the flexibility of defining most styles in code

**Comment by reececomo** [comment]
(i.e. save it to a `Blob`?)

**Comment by bigtimebuddy** [comment]
I'm confused by this. On-the-fly BitmapFont objects are cached when you create them. Maybe I don't understand what you're asking?

**Comment by reececomo** [comment]
Fair, I thought it would be more obvious given the context provided above (preloading/loading files), but I can see why "caching" is a dangerous word around here 😆 

I'm talking about caching the generated texture in the browser across browser page requests. As you know, when we provide an image or .fnt file those are cached on the client (checked against a backend, and skipped on a 304 Not Modified header). This would be some mechanism to save the texture/not generate it again if unchanged.

Same question/feature request could apply for any large generated textures that should ideally be persisted semi-permanently.

<hr />

alternative would be something we could run programmatically during webpack?

```
// fonts.js
const myFontStyle = new PIXI.BitmapFont(/* ... */);
myFontStyle.save('MyFont1.fnt');
```

No major issue if not - just trying to avoid manage ~20 .fnt files x 6 language scripts (latin, cyrillic, arabic, etc.)

**Comment by reececomo** [comment]
I'm talking ...dump it in localStorage kind of territory.

**Comment by bigtimebuddy** [comment]
Oh, that makes much more sense. Yeah, a dump in localStorage is probably it though there's a 5mb limit you'll likely bump up against. 20 fonts, depending on the size might approach that limit. Pixi doesn't have an easy way to extract the font, so you might need to add something there. Another approach is to use `@pixi/node` to generate and cache fonts server-side. Which could give you the right mix of on-demand but easy-ish request.

**Comment by reececomo** [comment]
Nice!

> Pixi doesn't have an easy way to extract the font, so you might need to add something there.

I guess that's the main bit - if only there was some fast way to dump the texture + "spritesheet json" to a .fnt file (or as a json/blob)

**Comment by bigtimebuddy** [comment]
Here's an example @reececomo: https://jsfiddle.net/bigtimebuddy/jzL5814f/

Essentially, you take all the data from the BitmapFont object and convert it into an XML that could be used later with `BitampFont.install`.

Hope that helps.

I'm going to close this issue as this is likely not something we are going to add to Pixi. I suggest you develop it as a plugin for the handful of folks that might want to export their dynamic fonts.

**Comment by reececomo** [comment]
Rad, thanks for the example!

**Comment by reececomo** [comment]
Thanks again for the example @bigtimebuddy, it's looking promising for bulk exports (across languages!).

Running into two issues when trying to load back the generated `.fnt` files, it's failing to load the base64'd data.

### Steps to reproduce:

Example file: [**example.fnt**](https://github.com/pixijs/pixijs/files/14692543/example.fnt.zip)

```ts
Assets.load('assets/font/bitmap/example.fnt'); // paraphrasing, but its roughly this
```

### What's going wrong?

- For some reason the .fnt files try to load assets relative to the .fnt file: "http://localhost:8089/assets/font/bitmap/<DATA_BLOB>"
- When I implemented a custom loader, it fails to load the base64'd data url.

### ❌ Custom loader:

I thought this might be solved with a custom loader, but the base64'd data that the custom loader receives is different than what's exported/in the file. URL encoding maybe?

```ts
const customLoader = {
  extension: ExtensionType.LoadParser,
  name: 'myFontDataBlobLoader',
  test: (url: string) => url.includes('assets/font/bitmap/data:image/png;base64'),
  async load(src: string) {
    return new Promise((resolve, reject) => {
      const dataUrl = src.split('assets/font/bitmap/')[1]!; // Get just the data url

      const img = new Image();
      img.crossOrigin = 'anonymous';
      img.src = dataUrl;
      img.onerror = reject;
      img.onload = () => resolve(Texture.from(img));
    });
  },
};

extensions.add(customLoader);
```

Apologies if this is really obvious, we're hitting niche territory

**Reply by bigtimebuddy** [reply]
I would suggest that you don't use Assets to load the font, but instead use this example:
https://jsfiddle.net/bigtimebuddy/jzL5814f/

Essentially, you want to use `BitmapFont.install` to reconstitute the font from the XML string. You need to do a little work to load the page textures from the font, but you can do this with Assets.

**Reply by reececomo** [reply]
> You need to do a little work to load the page textures from the font, but you can do this with Assets

Fair, though thats actually the only step I was struggling with in the above.

We can probably jury-rig a custom .fnt loader from the built-in PIXI one that will enable data urls. Basically anything that can programmatically create browser-cachable fonts was the goal, whether that's through a browser hack, or just straight up generating the .fnt/.png files at webpack time (or some other manual build time step).

If we were going to store the plaintext XML, we would probably want to dump to something smaller (JSON or binary buffers) and then unpack to XML. XML over a server/cloudfront benefits from Brotli, but we'll blow the whole local storage budget instantly if we did more than 10 fonts styles 😆 

edit: Have popped this on hold for today, it was more of a background task, but if I don't have a crack tomorrow then it'll jump up in priority right before launch 😆

**Reply by reececomo** [reply]
p.s. thanks for all your help, do appreciate it 🙏

**Comment by reececomo** [comment]
Here is what is coming through as the `url` / `src` for the custom loader:

- [**url-example1.txt**](https://github.com/pixijs/pixijs/files/14692639/url-example1.txt)

<details>
<summary>Differences summary: "It starts the same, but it diverges about halfway in"</summary>

It starts the same, but it diverges about halfway in:

<img width="1113" alt="diff example" src="https://github.com/pixijs/pixijs/assets/7390781/5f4e26fa-6418-44d6-a7d1-f395439e79c5">

### Resulting image:

#### ✅ Original export:

<img width="1068" alt="Screenshot 2024-03-21 at 8 12 12 pm" src="https://github.com/pixijs/pixijs/assets/7390781/4732d5de-26ab-462d-9f92-d1d76aa77851">

#### ❌ URL received by the loader:

<img width="1011" alt="Screenshot 2024-03-21 at 8 12 50 pm" src="https://github.com/pixijs/pixijs/assets/7390781/0f6d1724-6096-49fe-98f7-2177f62a5ac7">

</details>

**Comment by reececomo** [comment]
Related? https://github.com/pixijs/pixijs/issues/5199#issuecomment-434323023
